lib/xandra/connection.ex

```diff
@@ -651,30 +648,30 @@ defmodule Xandra.Connection do
     {:keep_state, data}
   end
 
-  def connected(:cast, {:timed_out_id, stream_id}, %__MODULE__{} = data) do
+  # The caller is notifying the connection that the request (on stream_id) timed
+  # out on its side (that is, the caller reached its "after" in the receive block).
+  # We need to remove the stream ID from the in-flight requests but still keep
+  # track of it, because C* might still send us a response for that query (and when it
+  # does, we will throw it away and free the timed-out stream ID).
+  #
+  # C* does not support CANCELING queries, by the way, otherwise that's what we'd do here.
+  def connected(:cast, {:request_timed_out_at_caller, stream_id}, %__MODULE__{} = data) do
     data = update_in(data.in_flight_requests, &Map.delete(&1, stream_id))
-    data = update_in(data.timed_out_ids, &Map.put(&1, stream_id, DateTime.utc_now()))
-
+    data = put_in(data.timed_out_ids[stream_id], System.system_time(:millisecond))
     {:keep_state, data}
   end
 
-  def connected({:timeout, :flush_timed_out_stream_id}, :flush, %__MODULE__{} = data) do
-    now = DateTime.utc_now()
+  def connected({:timeout, :flush_timed_out_stream_ids}, _content, %__MODULE__{} = data) do
+    now = System.system_time(:millisecond)
 
-    data =
-      update_in(
-        data.timed_out_ids,
-        &Enum.reduce(&1, %{}, fn {stream_id, timestamp}, acc ->
-          if DateTime.diff(now, timestamp, :minute) > @timed_out_stream_id_timeout_minutes do
-            acc
-          else
-            Map.put(acc, stream_id, timestamp)
-          end
-        end)
-      )
+    new_timed_out_ids =
```

Comment:
With https://github.com/whatyouhide/xandra/pull/358/commits/4e2ac5ee7573a07de90d91996439009d51f53e45, I get the following stacktrace, just trying to execute a query:

```
** (exit) exited in: :gen_statem.call(#PID<0.933.0>, {:checkout_state_for_next_request, #Reference<0.0.121219.3000101197.1395982337.31962>}, :infinity)
    ** (EXIT) an exception was raised:
        ** (BadMapError) expected a map, got: []
            (erts 14.2.1) :erlang.is_map_key(21374, [])
            (xandra 0.18.1) lib/xandra/connection.ex:900: Xandra.Connection.random_free_stream_id/2
            (xandra 0.18.1) lib/xandra/connection.ex:600: Xandra.Connection.connected/3
            (stdlib 5.2) gen_statem.erl:1395: :gen_statem.loop_state_callback/11
            (stdlib 5.2) proc_lib.erl:241: :proc_lib.init_p_do_apply/3
    (stdlib 5.2) gen.erl:246: :gen.do_call/4
    (stdlib 5.2) gen_statem.erl:923: :gen_statem.call/3
    (xandra 0.18.1) lib/xandra/connection.ex:158: Xandra.Connection.execute/4
    (xandra 0.18.1) lib/xandra/retry_strategy.ex:309: Xandra.RetryStrategy.run_on_cluster/5
```

Reply:
Also, it seems that preparing a query doesn't work and fails with `:timeout`

Reply:
@harunzengin I fixed the bug and added a regression test for that, it wasn't the nicest test to add but I’m happy we have a test for this now 🙃 

As for the `prepare`, I’m not sure what you mean. I have a test in there that exercises `Xandra.prepare/3` and it's passing. Do you have an example of when preparing a query fails?

Reply:
I tested this out, the timeouts in prepared were probably related to the first case, it is resolved now.